### ISSUE TITLE
#491 broke morepath_batching and morepath_wiki

### ISSUE THREAD

**Issue by taschini** [morepath/morepath#499]
#491 broke two of the demo apps. Here is the summary of py.test: 

```
FAIL src/morepath_batching/morepath_batching/tests/test_batching.py::test_random_access_record
FAIL src/morepath_wiki/morepath_wiki/tests/test_wiki.py::test_noncamelcase_page
```

Limiting our attention to morepath_batching, before #491:

```shell
cd src/morepath
git checkout c835016
cd ../morepath_batching
python -m morepath_batching
```

When you ask for a non-existing user you get a 404 response:
```console
$ curl 'http://127.0.0.1:5000/101' 

404 Not Found

The resource could not be found.

```

After #491:

```shell
cd ../morepath
git checkout master
cd ../morepath_batching
python -m morepath_batching
```

You get:

```console
$ curl 'http://127.0.0.1:5000/101'
<!DOCTYPE html>
<html>
  <head>
    <title>Morepath batching demo</title>
  </head>
  <body>
    <table>
      <tr>
        <th>Name</th>
        <th>Email</th>
        <th>Address</th>
      </tr>
      
      <tr>
        <td><a href="http://127.0.0.1:5000/0">Mr. Gregory Blake</a></td>
        <td>bergrobert@gmail.com</td>
        <td>3365 Garcia Turnpike Suite 825
Leefurt, MT 51992-1251</td>
      </tr>
      
      <tr>
        <td><a href="http://127.0.0.1:5000/1">Alex Kim</a></td>
        <td>brettmartin@waller.biz</td>
        <td>36521 Robinson Ramp
Port Stephen, MO 04547</td>
      </tr>
....
```

Which is exactly what you'd get if you asked for the whole list:

```
$ diff -s <(curl -s 'http://127.0.0.1:5000/101') <(curl -s 'http://127.0.0.1:5000/')
Files /dev/fd/63 and /dev/fd/62 are identical
```

Similar behaviour can be seen in morepath_wiki.

**Comment by faassen** [comment]
Good catch. Looks like the test suite has some holes. I will sit down and debug this, but it may have to wait until Monday.

**Comment by faassen** [comment]
I debugged this a bit, but haven't devised a proper fix yet. The problem occurs when you go into a path function that returns ``None``. The segment is consumed in this case. If this results in no segments being left unconsumed, it tries to get the thing registered for the root. If it finds it, it will get the view for it and display it.

The problem is that it *shouldn't* consume the segment if the path function for it returns ``None``. In the old case this was solved by not storing the stack as ``request.unconsumed`` if the creation function returned ``None``, but since we modify the unconsumed directly now, we now need code to store the stack to this original state.

**Comment by faassen** [comment]
@taschini could you retry this with the wiki and the batching demo?

**Comment by taschini** [comment]
I confirm that the test suites of the wiki and batching demo (and now also that of Morepath ;-) ) now work.

**Comment by faassen** [comment]
I improved the fix in the light of #500.
